src/shared/cloudformation/activation.ts

```diff
@@ -48,15 +46,106 @@ export async function activate(extensionContext: vscode.ExtensionContext): Promi
                 getIdeProperties().codelenses
             )
         )
-        getLogger().error('Failed to activate template registry', e)
+        getLogger().error('Failed to activate template registry: %s', e)
         // This prevents us from breaking for any reason later if it fails to load. Since
         // Noop watcher is always empty, we will get back empty arrays with no issues.
         globals.templateRegistry = new NoopWatcher() as unknown as CloudFormationTemplateRegistry
     }
     // If setting it up worked, add it to subscriptions so it is cleaned up at exit
     extensionContext.subscriptions.push(
-        globals.templateRegistry,
         Commands.register('aws.cloudFormation.newTemplate', () => createStarterTemplateFile(false)),
         Commands.register('aws.sam.newTemplate', () => createStarterTemplateFile(true))
     )
 }
+
+// True if template registry instance has been subscribed to disposables
+let isSubscribed = false
+
+// The current instance of the message which indicates the registry is being setup
+let setupProgressMessage: Thenable<void> | undefined
+
+/**
+ * Sets the `templateRegistry` property in the `globals` variable.
+ *
+ * This does the following:
+ * - Shows a 'loading' message to the user if the template registry variable is
+ *   requested but not yet available. For example, this would be triggered if
+ *   template file is opened.
+ * - Returns the correct registry object once all background loading is completed.
+ */
+function setTemplateRegistryInGlobals(registry: CloudFormationTemplateRegistry, disposables: vscode.Disposable[]) {
+    Object.defineProperty(globals, 'templateRegistry', {
+        set(newInstance: CloudFormationTemplateRegistry) {
+            this.instance = newInstance
```

Comment:
Yeah the logic should probably be added using the `static` keyword to `CloudFormationTemplateRegistry` instead of patching stuff. `this` refers to `globals` here.